sympy/sets/sets.py

```diff
@@ -1014,18 +1029,33 @@ def __new__(cls, *args, **kwargs):
         evaluate = kwargs.get('evaluate', global_evaluate[0])
 
         # flatten inputs to merge intersections and iterables
-        args = list(args)
 
+        # If the only arg is a list or another iterable,
+        # don't put the args tuple inside another list, instead
+        # put the elemnet of the tuple inside list
+        if len(args)==1:
+            args = args[0]
+            if iterable(args) and not isinstance(args,Basic):
+                args = list(args)
+            else: # if argument is an instance of Basic, don't convert to list, instead put inside a list
+                args = [args]
+        else:
+            args = list(args)
+
+        # flatten takes an arg and returns [arg] after applying changes
+        # to sets, Sets and iterables
         def flatten(arg):
+            if isinstance(arg, set):
+                return [_sympify(arg)]
             if isinstance(arg, Set):
                 if arg.is_Union:
                     return sum(map(flatten, arg.args), [])
                 else:
                     return [arg]
             if iterable(arg):  # and not isinstance(arg, Set) (implicit)
-                return sum(map(flatten, arg), [])
+                return [FiniteSet(*sum(map(flatten, arg), []))]
```

Comment:
Here, `sum` is used to merge lists. @asmeurer Is this situation different from the one where you said using `sum` gives quadratic performance, probable because of `map`?

Reply:
Using `sum` to merge lists always has quadratic performance, because it is essentially computing `(((list1 + list2) + list3) + list4) + ...)`. Each time a new list is created it has to allocate memory for the list and put the elements in it. On the other hand, using something like `[j for i in a for j in i]` only creates one list (Python uses doubling for allocation, so this is not an issue).

You can easily test this yourself if you do something like `a = [[1, 2] for i in range(10000)]` and compare the runtimes of `sum(a, [])` vs. `[j for i in a for j in i]`. For me it's 248 ms vs. 1.57 ms. And it becomes more drastic if you increase the number of lists. For 100000 lists, it's 27.6 s vs. 12.4 ms (a 2000x speed difference). 

Actually, this same sort of issue can also occur with SymPy expressions, which is why you should use `Add(*args)` instead of `sum(args)`. `sum` should only be used for numbers (integers and floats). The reason this issue doesn't happen with integers is because integers do not grow linearly in size with summation (adding `n` integers does not create an integer that requires `n` times as much memory to represent). 

Anyway, I thought we were removing flattening, so I'm unclear why this is still here. Is there still logic in Union that fully flattens its arguments?

Reply:
No, There is no flattening. The name of the function was flatten originally and I didn't change it. This is required logic to convert arguments such as a Union, python Set, into a list. Basically, processing arguments and converting in a standard format.

Would changing the name of function to something else be better?

Reply:
I understood the complexity issues with using `sum`. But, I was wondering why was it used in the first place if it is not optimal, here?